File: examples/footguns/07_causal_ordering/main.go
// Footgun: Causal Message Ordering — Go
//
// Go channels are FIFO per sender, but there is no causal ordering across
// independent channels. If A sends on ch1 and independently B sends on ch2,
// a receiver reading from both channels cannot determine which send
// happened first without out-of-band coordination.
//
// The practical consequence: a relay pattern where B forwards A's signal
// to C does NOT automatically carry A's writes to C. C may read shared
// state that A wrote before the relay, but see it as stale.
//
// Go's memory model (revised 2022) is explicit: a channel send
// happens-before the corresponding receive, but only for the same channel.
// Crossing channel boundaries breaks the happens-before chain unless you
// synchronize explicitly (e.g. via a mutex or a single channel for both
// the data and the signal).

package main

import (
	"fmt"
	"sync"
)

// BROKEN: relay pattern with split channels. The happens-before chain
// from writer → relay → reader is not preserved for sharedData.
func brokenRelay() {
	var sharedData string
	var mu sync.Mutex

	ch1 := make(chan struct{}, 1) // writer → relay
	ch2 := make(chan struct{}, 1) // relay  → reader

	var wg sync.WaitGroup
	wg.Add(3)

	// Writer: sets data, then signals relay.
	go func() {
		defer wg.Done()
		mu.Lock()
		sharedData = "important result"
		mu.Unlock()
		ch1 <- struct{}{} // signals relay
	}()

	// Relay: receives from ch1, forwards to ch2.
	// The happens-before from ch1 receive covers the writer's signal,
	// but the relay does NOT re-establish happens-before for sharedData
	// when it sends on ch2. ch2 receive only happens-before ch2 send.
	go func() {
		defer wg.Done()
		<-ch1
		ch2 <- struct{}{} // forward — does NOT carry sharedData h-b
	}()

	// Reader: receives "go" from relay, reads sharedData.
	go func() {
		defer wg.Done()
		<-ch2
		// Go memory model: ch2 receive h-b ch2 send (relay's side).
		// But sharedData's write h-b ch1 send, not ch2 send.
		// There is no direct happens-before between sharedData write
		// and ch2 receive. On real hardware this is usually fine due
		// to OS scheduler ordering, but it is not guaranteed by the
		// memory model.
		mu.Lock()
		fmt.Printf("reader saw: '%s'\n", sharedData)
		mu.Unlock()
	}()

	wg.Wait()
}

// CORRECT: send the data itself through the channel, not a separate signal.
// The channel carry the happens-before; no split between data and signal.
func correctRelay() {
	ch := make(chan string, 1) // carries both data and h-b guarantee

	var wg sync.WaitGroup
	wg.Add(2)

	go func() {
		defer wg.Done()
		ch <- "important result" // data and signal in one send
	}()

	go func() {
		defer wg.Done()
		data := <-ch // h-b guaranteed: write happens-before this receive
		fmt.Printf("reader saw: '%s'\n", data)
	}()

	wg.Wait()
}

func main() {
	fmt.Println("--- broken relay (split channel + shared memory) ---")
	brokenRelay()

	fmt.Println("--- correct relay (data through channel) ---")
	correctRelay()
}

File: examples/footguns/07_causal_ordering/main.rs
// Footgun: Causal Message Ordering — Rust
//
// Rust's std::sync::mpsc channels are FIFO per sender and establish
// happens-before between send and receive on the same channel. But
// crossing channel boundaries in a relay pattern breaks the chain
// for any out-of-band shared state, exactly as in Go.
//
// Rust prevents data races (no unsynchronized access to sharedData),
// but it does not enforce causal ordering of messages across channels.
// The programmer must structure the code so that the data itself travels
// through the channel, not a separate signal about the data.

use std::sync::{mpsc, Arc, Mutex};
use std::thread;

// BROKEN: writer updates shared state, signals relay via ch1.
// Relay forwards signal via ch2. Reader checks shared state after ch2.
// The happens-before chain for shared_data is broken at the relay hop.
fn broken_relay() {
    let shared_data: Arc<Mutex<String>> = Arc::new(Mutex::new(String::new()));
    let (tx1, rx1) = mpsc::channel::<()>(); // writer → relay
    let (tx2, rx2) = mpsc::channel::<()>(); // relay  → reader

    let data_w = Arc::clone(&shared_data);
    let writer = thread::spawn(move || {
        *data_w.lock().unwrap() = "important result".to_string();
        tx1.send(()).unwrap(); // signal relay
    });

    let relay = thread::spawn(move || {
        rx1.recv().unwrap();
        // rx1 receive h-b tx1 send (writer's signal).
        // But tx2 send does NOT transitively carry shared_data's h-b to
        // the reader. The reader's rx2 receive only h-b tx2 send here.
        tx2.send(()).unwrap();
    });

    let data_r = Arc::clone(&shared_data);
    let reader = thread::spawn(move || {
        rx2.recv().unwrap();
        // Mutex acquire provides a barrier, so in practice this is safe
        // here because the Mutex itself provides sequential consistency.
        // However, if shared_data were an atomic or raw pointer instead,
        // the happens-before chain would be broken and we could observe
        // the pre-write value on weakly-ordered hardware.
        println!("broken relay: '{}'", data_r.lock().unwrap());
    });

    writer.join().unwrap();
    relay.join().unwrap();
    reader.join().unwrap();
}

// CORRECT: send the data through the channel; the channel carries
// both the value and the happens-before guarantee in one operation.
fn correct_relay() {
    let (tx1, rx1) = mpsc::channel::<String>(); // carries the data
    let (tx2, rx2) = mpsc::channel::<String>(); // carries the data

    let writer = thread::spawn(move || {
        tx1.send("important result".to_string()).unwrap();
    });

    let relay = thread::spawn(move || {
        let data = rx1.recv().unwrap(); // h-b: data is visible here
        tx2.send(data).unwrap();        // forward the data itself
    });

    let reader = thread::spawn(move || {
        let data = rx2.recv().unwrap(); // h-b: data guaranteed visible
        println!("correct relay: '{}'", data);
    });

    writer.join().unwrap();
    relay.join().unwrap();
    reader.join().unwrap();
}

fn main() {
    println!("--- broken relay (signal separate from data) ---");
    broken_relay();

    println!("--- correct relay (data through channel) ---");
    correct_relay();
}

// Compile: rustc main.rs -o causal && ./causal
//
// Key insight: happens-before is a property of specific synchronization
// operations, not of logical program flow. A relay that receives on ch1
// and sends on ch2 does not automatically propagate the h-b from ch1
// to the reader of ch2 unless the data itself moves through ch2.
